src/lib/grammar.rs

```diff
@@ -1,166 +1,84 @@
-use std::collections::{HashMap, HashSet};
-use std::fmt;
+use std::collections::HashMap;
 
-pub struct Grammar {
-    pub start: String,
-    pub rules: HashMap<String, Rule>,
-    pub tokens: HashSet<String>
-}
+use grammar_ast;
+
+/// A type specifically for nonterminal indices (i.e. corresponding to a rule
+/// name).
+pub type NIdx = i32;
```

Comment:
On a 64 bit machine, usize will be 64 bits. The enum then adds tagging bits, which means each symbol would probably take up 128 bits in memory. By making these types 32 bit, we should happily squeeze into 64 bits. And I reckon there won't be too many grammars with ~4294967296 nonterminals or ~4294967296 terminals :)